### PR TITLE
when creating self group-reshare do not share to self

### PR DESCRIPTION
**Pull request by mrow4a**
fixes https://github.com/owncloud/enterprise/issues/4382
related https://github.com/owncloud/core/pull/38542

This PR solved the issue where user is resharing with group that himself belongs to, and where further resharing or increasing permissions is not possible due to self-share with reduced permission

- [x] manual tests
- [x] add unit and integration tests

### CODE REVIEW

File: tests/acceptance/features/webUISharingInternalGroups2/shareWithGroupsEdgeCases.feature [hunk 1]
```diff
@@ -349,3 +350,55 @@ Feature: Sharing files and folders with internal groups
       | uid_owner   | Carol          |
       | share_with  | grp2           |
       | permissions | 7              |
+
+  @skipOnOcV10.3 @skipOnOcV10.4 @skipOnOcV10.5 @skipOnOcV10.6
+  Scenario: Reshare with group that user is member of should not create mount in the root folder of resherer
```
Comment (phil-davis) [inline tests/acceptance/features/webUISharingInternalGroups2/shareWithGroupsEdgeCases.feature:355]:
```suggestion
  Scenario: Reshare with group that user is member of should not create mount in the root folder of resharer
```

File: tests/acceptance/features/webUISharingInternalGroups2/shareWithGroupsEdgeCases.feature [hunk 2]
```diff
@@ -349,3 +350,55 @@ Feature: Sharing files and folders with internal groups
...
+      | groupname |
+      | grp1      |
+      | grp2      |
+    And user "Alice" has been added to group "grp1"
+    And user "Carol" has been added to group "grp1"
+    And user "Alice" has been added to group "grp2"
+    And user "Brian" has been added to group "grp2"
+    And user "Carol" has shared folder "/simple-folder" with group "grp1"
+    And user "Alice" has logged in using the webUI
+    When the user shares file "simple-folder" with group "grp2" using the webUI without closing the share dialog
```
Comment (phil-davis) [inline tests/acceptance/features/webUISharingInternalGroups2/shareWithGroupsEdgeCases.feature:366]:
```suggestion
    When the user shares folder "simple-folder" with group "grp2" using the webUI without closing the share dialog
```

It does not make a difference to the UI action. Changing `file` to `folder` here just makes it more readable.

Why do you want `without closing the share dialog` ?
I think that makes the UI interactions stop a bit early.

File: tests/acceptance/features/webUISharingInternalGroups2/shareWithGroupsEdgeCases.feature [hunk 3]
```diff
@@ -349,3 +350,55 @@ Feature: Sharing files and folders with internal groups
...
+    And the user opens folder "simple-folder" using the webUI
+    And the user shares file "simple-empty-folder" with group "grp2" using the webUI without closing the share dialog
```
Comment (phil-davis) [inline tests/acceptance/features/webUISharingInternalGroups2/shareWithGroupsEdgeCases.feature:368]:
```suggestion
    And the user shares folder "simple-empty-folder" with group "grp2" using the webUI without closing the share dialog
```

Why do you want `without closing the share dialog` ?
I think that makes the UI interactions stop a bit early.

File: tests/acceptance/features/webUISharingInternalGroups2/shareWithGroupsEdgeCases.feature [hunk 4]
```diff
@@ -349,3 +350,55 @@ Feature: Sharing files and folders with internal groups
...
+      | groupname |
+      | grp1      |
+      | grp2      |
+    And user "Alice" has been added to group "grp1"
+    And user "Carol" has been added to group "grp1"
+    And user "Alice" has been added to group "grp2"
+    And user "Brian" has been added to group "grp2"
+    And user "Carol" has shared folder "/simple-folder" with group "grp1"
+    And user "Alice" has logged in using the webUI
+    When the user shares file "simple-folder" with group "grp2" using the webUI without closing the share dialog
```
Comment (phil-davis) [inline tests/acceptance/features/webUISharingInternalGroups2/shareWithGroupsEdgeCases.feature:387]:
```suggestion
    When the user shares folder "simple-folder" with group "grp2" using the webUI without closing the share dialog
```

File: tests/acceptance/features/webUISharingInternalGroups2/shareWithGroupsEdgeCases.feature [hunk 5]
```diff
@@ -349,3 +350,55 @@ Feature: Sharing files and folders with internal groups
       | uid_owner   | Carol          |
       | share_with  | grp2           |
       | permissions | 7              |
+
+  @skipOnOcV10.3 @skipOnOcV10.4 @skipOnOcV10.5 @skipOnOcV10.6
```
Comment (phil-davis) [inline tests/acceptance/features/webUISharingInternalGroups2/shareWithGroupsEdgeCases.feature:354]:
```suggestion
  @skipOnOcV10.3 @skipOnOcV10.4 @skipOnOcV10.5 @skipOnOcV10.6 @skipOnOcV10.7.0
```

File: tests/acceptance/features/webUISharingInternalGroups2/shareWithGroupsEdgeCases.feature [hunk 6]
```diff
@@ -349,3 +350,55 @@ Feature: Sharing files and folders with internal groups
...
+    When the user shares file "simple-folder" with group "grp2" using the webUI without closing the share dialog
+    And the user opens folder "simple-folder" using the webUI
+    And the user shares file "simple-empty-folder" with group "grp2" using the webUI without closing the share dialog
+    Then as "Alice" folder "/simple-folder" should exist
+    And user "Alice" should be able to upload file "filesForUpload/textfile.txt" to "simple-folder/textfile.txt"
+    And as "Alice" folder "/simple-empty-folder" should not exist
+    When user "Alice" has created folder "/simple-empty-folder"
+    Then as "Alice" folder "/simple-empty-folder" should exist
+
+  @skipOnOcV10.3 @skipOnOcV10.4 @skipOnOcV10.5 @skipOnOcV10.6
```
Comment (phil-davis) [inline tests/acceptance/features/webUISharingInternalGroups2/shareWithGroupsEdgeCases.feature:375]:
```suggestion
  @skipOnOcV10.3 @skipOnOcV10.4 @skipOnOcV10.5 @skipOnOcV10.6 @skipOnOcV10.7.0
```

**Review by phil-davis — COMMENTED** [review]
Note: probably the test scenarios could be just with the API. But I guess it is also useful that the test scenarios are driving the UI - that verifies that the UI is also correctly rendering the various situations.

**Comment by jvillafanez** [comment]
Code looks good :+1: . Waiting for movement on the acceptance test's comments

**Review by phil-davis — APPROVED** [review]
I refactored some steps at the end of a scenario to be:
`And user "Alice" should be able to create folder "/simple-empty-folder"`

We have other "Then user Alice should be able to do something" steps. This one will also be useful to have (rather than writing and extra When/Then section at the end of a scenario.

**Pull request merged**